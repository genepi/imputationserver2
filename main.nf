#!/usr/bin/env nextflow
/*
========================================================================================
    genepi/nf-imputationserver
========================================================================================
    Github : https://github.com/genepi/nf-imputationserver
    Author: Lukas Forer & Sebastian SchÃ¶nherr
    ---------------------------
*/

if (params.refpanel_yaml){
    params.refpanel = RefPanelUtil.loadFromFile(params.refpanel_yaml)
}

requiredParams = [
    'project', 'files', 'output', 'refpanel'
]

for (param in requiredParams) {
    if (params[param] == null) {
      exit 1, "Parameter ${param} is required."
    }
}

//TODO create json validation file
def phasing_engine = params.phasing.engine

if (phasing_engine != 'eagle' && phasing_engine != 'beagle' && phasing_engine != 'no_phasing' ) {
    println "::error:: For phasing, only options 'eagle', 'beagle' or 'no_phasing' are allowed."
    exit 1
}


// create random password when not set by user
if (params.password == null) {
    params.encryption_password = PasswordCreator.createPassword()   
} else {
    params.encryption_password = params.password
}


Channel
    .fromPath(params.files)
    .set {files}

files.ifEmpty {
    println "::error:: No vcf.gz input files detected."
    exit 1
}

// Find site files from full pattern and make site file pattern relative
params.refpanel.sites_pattern = "${params.refpanel.sites}"
params.refpanel.sites = "./${file(params.refpanel.sites).fileName}"

site_files_ch = Channel.of(1..22, 'X', 'MT')
    .map {
        it -> 
            def sites_file = file(PatternUtil.parse(params.refpanel.sites_pattern, [chr: it]))
            def sites_file_index = file(PatternUtil.parse(params.refpanel.sites_pattern+ ".tbi", [chr: it]))
            
            if(!sites_file.exists()){
                return null;
            }  

            if(sites_file.exists() && !sites_file_index.exists()){
                error "Missing tabix index for " + sites_file
            }  
        
            return tuple(sites_file, sites_file_index); 
    }

include { INPUT_VALIDATION } from './workflows/input_validation'
include { QUALITY_CONTROL } from './workflows/quality_control'
include { PHASING } from './workflows/phasing'
include { IMPUTATION } from './workflows/imputation'
include { ENCRYPTION } from './workflows/encryption'
include { ANCESTRY_ESTIMATION } from './workflows/ancestry_estimation'
include { PGS_CALCULATION } from './workflows/pgs_calculation'

workflow {

    println "Welcome to ${params.service.name} (${workflow.manifest.version})"

    if (params.imputation.enabled){

        INPUT_VALIDATION()

        QUALITY_CONTROL(
            INPUT_VALIDATION.out,
            site_files_ch.collect()
        )

        if (params.mode == 'imputation') {

            phased_ch =  QUALITY_CONTROL.out.qc_metafiles

            if (phasing_engine != 'no_phasing') { 

                PHASING(
                    QUALITY_CONTROL.out.qc_metafiles
                )

                phased_ch = PHASING.out.phased_ch

            }
                 
            IMPUTATION(
                phased_ch
            )
            
            if (params.merge_results === true) {
                ENCRYPTION(
                    IMPUTATION.out.groupTuple()
                )
            }
            
        }
    }
    
    // handles empty objects (e.g. cloudgene)
    ancestry_enabled = params.ancestry != null && params.ancestry != "" && params.ancestry.enabled

    if (ancestry_enabled) {
        ANCESTRY_ESTIMATION()
    }

    if (params.pgs.enabled) {

        PGS_CALCULATION(
            IMPUTATION.out,
            ancestry_enabled ? ANCESTRY_ESTIMATION.out : Channel.empty()
        )
        
    }
    
}

workflow.onComplete {
    //TODO: use templates
    //TODO: move in EmailHelper class
    //see https://www.nextflow.io/docs/latest/mail.html for configuration etc...
    // Nfcore Template: https://github.com/nf-core/rnaseq/blob/b89fac32650aacc86fcda9ee77e00612a1d77066/lib/NfcoreTemplate.groovy#L155
   
    if (!workflow.success) {
        if (params.send_mail && params.user.email != null){
            sendMail{
                to "${params.user.email}"
                subject "[${params.service.name}] Job ${params.project} failed."
                body "Dear ${params.user.name}, \n Your job has failed.\n\n More details about the errors can be found at the following link: ${params.service.url}/index.html#!jobs/${params.project}"
            }
        }
        println "::error:: Imputation failed." 
        return
    }

    //submit counters on success
	println "::submit-counter name=samples::"
    println "::submit-counter name=genotypes::"
    println "::submit-counter name=chromosomes::"
    println "::submit-counter name=runs::"

    println "::set-value-and-submit name=reference_panel::${params.refpanel.id}"
    println "::set-value-and-submit name=phasing_engine::${phasing_engine}"
    println "::set-value-and-submit name=genome_build::${params.build}"


    // imputation job
    if (params.merge_results === true && params.encryption.enabled === true) {

        if (params.send_mail && params.user.email != null) {
            sendMail{
                to "${params.user.email}"
                subject "[${params.service.name}] Job ${params.project} is complete."
                body "Dear ${params.user.name}, \n Your imputation job has finished succesfully. The password for the imputation results is: ${params.encryption_password}\n\n You can download the results from the following link: ${params.service.url}/index.html#!jobs/${params.project}"
            }
            println "::message:: Data have been exported successfully. We have sent a notification email to <b>${params.user.email}</b>"
        } else {
            println "::message:: Data have been exported successfully. We encrypted the results with the following password: <b>${params.encryption_password}</b>"
        }
        return
    }

    //PGS job
    if (params.send_mail && params.user.email != null) {
        sendMail{
            to "${params.user.email}"
            subject "[${params.service.name}] Job ${params.project} is complete."
            body "Dear ${params.user.name}, \n Your PGS job has finished successfully. \n\n You can download the results from the following link: ${params.service.url}/index.html#!jobs/${params.project}"
        }
        println "::message:: Data have been exported successfully. We have sent a notification email to <b>${params.user.email}</b>"
    } else {
        println "::message:: Data have been exported successfully."
    }
 
}