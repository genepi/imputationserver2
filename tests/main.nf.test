@Grab(group='net.lingala.zip4j', module='zip4j', version='2.11.5')
import net.lingala.zip4j.ZipFile;

nextflow_pipeline {

    def PASSWORD = "test-password"
	def TOTAL_REFPANEL_CHR20_B37 = 63402;
	def TOTAL_REFPANEL_CHR20_B38 = 63384;
	def ONLY_IN_INPUT = 78;
	def TOTAL_SNPS_INPUT = 7824;
	def SNPS_MONOMORPHIC = 11;

    name "Test Workflow main.nf"
    script "main.nf"

    test("Should run QC-only") {

        when {
            params {
                project  = "test-job"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                population = "eur"
                mode = "qc-only"
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
            }   
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")
       
        }

    }

    test("Should run QC-only and check statistic files") {

        when {
            params {
                project  = "test-job"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                population = "eur"
                mode = "qc-only"
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")
            def typed_file = file("${outputDir}/statistics/typed-only.txt")
            assert typed_file.exists()
            def snps_file = file("${outputDir}/statistics/snps-excluded.txt")
            assert snps_file.exists()
       
        }

    }
    
    test("Should run with phased data") {

        when {
            params {
                project  = "testPipelineWithPhased"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                population = "eur"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
                phasing = "no_phasing"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");
            def file = path("${outputDir}/chr20.dose.vcf.gz").vcf
            assert file.getChromosome() == "20"
            assert file.getNoSamples() == 51;
            assert file.isPhased()
            assert file.getNoSnps() == TOTAL_REFPANEL_CHR20_B37 + ONLY_IN_INPUT
            
            //check correct number of snps in info.gz file
            assert path("${outputDir}/chr20.info.gz").linesGzip.size() == 1 + TOTAL_REFPANEL_CHR20_B37 + ONLY_IN_INPUT

        }

    }

    test("Should run with phased data and meta option") {

        when {
            params {
                project  = "testPipelineWithPhasedAndMetaOption"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                population = "eur"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
                phasing = "no_phasing"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");
            def file = path("${outputDir}/chr20.dose.vcf.gz").vcf
            def fileMeta = path("${outputDir}/chr20.empiricalDose.vcf.gz").vcf
            assert fileMeta.getNoSnps() == 7735
            assert file.getChromosome() == "20"
            assert file.getNoSamples() == 51;
            assert file.isPhased()
            assert file.getNoSnps() == TOTAL_REFPANEL_CHR20_B37 + ONLY_IN_INPUT
            
            //check correct number of snps in info.gz file
            assert path("${outputDir}/chr20.info.gz").linesGzip.size() == 1 + TOTAL_REFPANEL_CHR20_B37 + ONLY_IN_INPUT

        }

    }

    test("Should fail with phased data and empty phasing option") {

        when {
            params {
                project  = "testPipelineWithPhasedAndEmptyPhasing"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                population = "eur"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
                phasing = ""
            }
        }

        then {
            assert workflow.failed

        }

    }
    
    // this testcould be deleted later, since now phasing must be actively deactivated
    test("Should run with phased data and no phasing selected") {

        when {
            params {
                project  = "testPipelineWithPhasedAndNoPhasingSelected"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                population = "eur"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
                phasing = "no_phasing"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");
            def file = path("${outputDir}/chr20.dose.vcf.gz").vcf
            assert file.getChromosome() == "20"
            assert file.getNoSamples() == 51;
            assert file.isPhased()
            assert file.getNoSnps() == TOTAL_REFPANEL_CHR20_B37 + ONLY_IN_INPUT
            
            //check correct number of snps in info.gz file
            assert path("${outputDir}/chr20.info.gz").linesGzip.size() == 1 + TOTAL_REFPANEL_CHR20_B37 + ONLY_IN_INPUT

        }

    }    

    test("Should run with eagle") {

        when {
            params {
                project  = "testPipelineWithEagle"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                population = "eur"
                phasing = "no_phasing"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
                phasing = "eagle"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");

            def file = path("${outputDir}/chr20.dose.vcf.gz").vcf
            assert file.getChromosome() == "20"
            assert file.getNoSamples() == 51;
            assert file.isPhased()
            assert file.getNoSnps() == TOTAL_REFPANEL_CHR20_B37

            //check correct number of snps in info.gz file
            assert path("${outputDir}/chr20.info.gz").linesGzip.size() == 1 + TOTAL_REFPANEL_CHR20_B37

        }

    }

    test("Should run with eagle and validate header") {

        when {
            params {
                project  = "testValidatePanelWithEagle"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                population = "eur"
                phasing = "no_phasing"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");

            def header = getVcfHeader(path("${outputDir}/chr20.dose.vcf.gz"))
            assert header.getOtherHeaderLine("mis_phasing").getValue() == params.phasing
            assert header.getOtherHeaderLine("mis_pipeline").getKey()
            assert header.getOtherHeaderLine("mis_panel").getKey()
        }

    }

    test("Should run with eagle and r2 filter") {

        when {
            params {
                project  = "testPipelineWithEagleAndR2Filter"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*.vcf.gz"
                population = "eur"
                phasing = "no_phasing"
                password = PASSWORD
                r2Filter = 0.5
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");

            def file = path("${outputDir}/chr20.dose.vcf.gz").vcf
            assert file.getChromosome() == "20"
            assert file.getNoSamples() == 51
            assert file.isPhased()

            def lines = path("${outputDir}/chr20.info.gz").linesGzip.size() - 1
            assert file.getNoSnps() == lines

            assert TOTAL_REFPANEL_CHR20_B37 > file.getNoSnps()
          
        }

    }

    test("Should run with phased data from hg19 to hg38") {

        when {
            params {
                project  = "testPipelineWithEagleHg38ToHg19"
                build = "hg19"
                files = "$projectDir/tests/data/input/chr20-phased/*vcf.gz"
                population = "eur"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2-hg38/cloudgene.yaml"
                output = "${outputDir}"
                phasing = "no_phasing"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");

            def file = path("${outputDir}/chr20.dose.vcf.gz").vcf
            assert file.getChromosome() == "20"
            assert file.getNoSamples() == 51
            assert file.isPhased()

            assert TOTAL_REFPANEL_CHR20_B38 + ONLY_IN_INPUT == file.getNoSnps()
          
        }

    }

    test("Should run with eagle from hg19 to hg38") {

        when {
            params {
                project  = "testPipelineWithEagleHg19ToHg38"
                files = "$projectDir/tests/data/input/chr20-unphased/*vcf.gz"
                population = "eur"
                build = "hg19"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2-hg38/cloudgene.yaml"
                output = "${outputDir}"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");

            def file = path("${outputDir}/chr20.dose.vcf.gz").vcf
            assert file.getChromosome() == "20"
            assert file.getNoSamples() == 51
            assert file.isPhased()

            assert TOTAL_REFPANEL_CHR20_B38 == file.getNoSnps()
          
        }

    }

    test("Should run with eagle from hg38 to hg38") {

        when {
            params {
                project  = "testPipelineWithEagleHg38ToHg38"
                build = "hg38"
                files = "$projectDir/tests/data/input/chr20-unphased-hg38/*vcf.gz"
                population = "eur"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2-hg38/cloudgene.yaml"
                output = "${outputDir}"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");

            def file = path("${outputDir}/chr20.dose.vcf.gz").vcf
            assert file.getChromosome() == "20"
            assert file.getNoSamples() == 51
            assert file.isPhased()

            assert TOTAL_REFPANEL_CHR20_B38 == file.getNoSnps()
          
        }

    }
    
    test("Should run without phasing from hg38 to hg19") {

        when {
            params {
                project  = "testPipelineWithPhasedHg38ToHg19"
                build = "hg38"
                files = "$projectDir/tests/data/input/chr20-phased-hg38/*vcf.gz"
                population = "eur"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
                phasing = "no_phasing"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");

            def file = path("${outputDir}/chr20.dose.vcf.gz").vcf
            assert file.getChromosome() == "20"
            assert file.getNoSamples() == 51
            assert file.isPhased()

            assert TOTAL_REFPANEL_CHR20_B37 + ONLY_IN_INPUT == file.getNoSnps()
          
        }

    }

    test("Should run with eagle from hg38 to hg19") {

        when {
            params {
                project  = "testPipelineWithEagleHg38ToHg19"
                build = "hg38"
                files = "$projectDir/tests/data/input/chr20-unphased-hg38/*.vcf.gz"
                population = "eur"
                password = PASSWORD
                refpanel_yaml = "$projectDir/tests/data/refpanels/hapmap2/cloudgene.yaml"
                output = "${outputDir}"
            }
        }

        then {
            assert workflow.success

            def quality_control_log = file("${outputDir}/cloudgene.report.json")
            assert quality_control_log.exists()
            assert quality_control_log.text.contains("Remaining sites in total: 7,735")

            def imputed_chr_20 = file("${outputDir}/chr_20.zip");
            assert imputed_chr_20.exists()
            ZipFile zipFile = new ZipFile(imputed_chr_20, PASSWORD.toCharArray());
            zipFile.extractAll("${outputDir}");

            def file = path("${outputDir}/chr20.dose.vcf.gz").vcf
            assert file.getChromosome() == "20"
            assert file.getNoSamples() == 51
            assert file.isPhased()

            assert TOTAL_REFPANEL_CHR20_B37 == file.getNoSnps()
          
        }

    }
}