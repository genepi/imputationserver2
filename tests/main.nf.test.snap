{
    "Should run with eagle and without encryption and compression": {
        "content": [
            "[{\"command\":\"LOG\",\"params\":[\"Reference Panel Ranges: genome-wide\"]},{\"command\":\"BEGIN_TASK\",\"params\":[\"Running Calculating QC Statistics...\"]},{\"command\":\"UPDATE_TASK\",\"params\":[\"Calculating QC Statistics [1/1]\\n\\nAnalyze file chr20.R50.merged.1.330k.recode.small.vcf.gz...\",3]},{\"command\":\"END_TASK\",\"params\":[\"Calculating QC Statistics\",0]},{\"command\":\"MESSAGE\",\"params\":[\"\\u003cb\\u003eStatistics:\\u003c/b\\u003e \\nAlternative allele frequency \\u003e 0.5 sites: 2,296\\nReference Overlap: 99.00 %\\nMatch: 7,735\\nAllele switch: 0\\nStrand flip: 0\\nStrand flip and allele switch: 0\\nA/T, C/G genotypes: 0\\n\\u003cb\\u003eFiltered sites:\\u003c/b\\u003e \\nFilter flag set: 0\\nInvalid alleles: 0\\nMultiallelic sites: 0\\nDuplicated sites: 0\\nNonSNP sites: 0\\nMonomorphic sites: 11\\nAllele mismatch: 0\\nSNPs call rate \\u003c 90%: 0\",0]},{\"command\":\"MESSAGE\",\"params\":[\"Excluded sites in total: 11\\nRemaining sites in total: 7,735\\nSee snps-excluded.txt for details\\nTyped only sites: 78\\nSee typed-only.txt for details\\n\",2]}]"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-06-15T14:38:57.145373"
    },
    "Should run QC-only": {
        "content": [
            "[{\"command\":\"LOG\",\"params\":[\"Reference Panel Ranges: genome-wide\"]},{\"command\":\"BEGIN_TASK\",\"params\":[\"Running Calculating QC Statistics...\"]},{\"command\":\"UPDATE_TASK\",\"params\":[\"Calculating QC Statistics [1/1]\\n\\nAnalyze file chr20.R50.merged.1.330k.recode.small.vcf.gz...\",3]},{\"command\":\"END_TASK\",\"params\":[\"Calculating QC Statistics\",0]},{\"command\":\"MESSAGE\",\"params\":[\"\\u003cb\\u003eStatistics:\\u003c/b\\u003e \\nAlternative allele frequency \\u003e 0.5 sites: 2,296\\nReference Overlap: 99.00 %\\nMatch: 7,735\\nAllele switch: 0\\nStrand flip: 0\\nStrand flip and allele switch: 0\\nA/T, C/G genotypes: 0\\n\\u003cb\\u003eFiltered sites:\\u003c/b\\u003e \\nFilter flag set: 0\\nInvalid alleles: 0\\nMultiallelic sites: 0\\nDuplicated sites: 0\\nNonSNP sites: 0\\nMonomorphic sites: 11\\nAllele mismatch: 0\\nSNPs call rate \\u003c 90%: 0\",0]},{\"command\":\"MESSAGE\",\"params\":[\"Excluded sites in total: 11\\nRemaining sites in total: 7,735\\nSee snps-excluded.txt for details\\nTyped only sites: 78\\nSee typed-only.txt for details\\n\",2]}]",
            "#Position\n20:161502:C:T\n20:775376:T:G\n20:775842:C:T\n20:1438230:C:T\n20:2638579:T:C\n20:3193842:C:A\n20:3580168:G:A\n20:3811433:T:C\n20:4063862:A:G\n20:4578080:G:A\n20:7541572:A:G\n20:7581461:T:G\n20:7928922:A:G\n20:8364799:G:T\n20:10243251:T:C\n20:11005238:A:C\n20:12173825:T:C\n20:13290738:T:C\n20:13582476:A:C\n20:13650883:T:G\n20:13929943:C:T\n20:13998098:G:A\n20:14794361:A:G\n20:14921606:G:A\n20:16066764:G:A\n20:16070941:C:T\n20:16116624:G:A\n20:16120763:A:G\n20:17596155:C:A\n20:17600357:G:A\n20:17845755:T:C\n20:18284404:C:T\n20:18389203:T:C\n20:19025636:C:T\n20:20130511:T:C\n20:20516020:T:C\n20:21628586:C:T\n20:21863397:A:G\n20:21939656:C:T\n20:23682810:A:G\n20:23725019:A:G\n20:24261606:T:C\n20:24691928:T:G\n20:24955981:T:C\n20:25136824:A:G\n20:25948327:A:G\n20:26307797:T:C\n20:31590686:C:A\n20:35673056:C:T\n20:39048873:A:C\n20:39792063:A:G\n20:40093956:A:C\n20:40387411:T:C\n20:40706295:T:C\n20:43506247:G:A\n20:43668473:G:A\n20:44011546:C:T\n20:46332216:T:C\n20:46435339:T:C\n20:47263751:T:C\n20:47479776:C:T\n20:48143362:T:C\n20:48403371:T:C\n20:50395283:T:C\n20:50475418:G:A\n20:53251910:G:A\n20:53515863:C:A\n20:54264333:C:T\n20:54692704:A:G\n20:54961463:T:C\n20:56637477:A:G\n20:57304273:C:T\n20:58098827:G:A\n20:58114167:T:C\n20:59121456:A:G\n20:61381218:G:A\n20:61644330:A:G\n20:61706750:C:T",
            "#Position\tFilterType\t Info\n20:2187430:G:A\tMonomorphic\n20:24404652:C:T\tMonomorphic\n20:24424220:C:T\tMonomorphic\n20:25282967:C:T\tMonomorphic\n20:39005139:C:T\tMonomorphic\n20:45867852:A:G\tMonomorphic\n20:45873923:C:T\tMonomorphic\n20:48575315:C:T\tMonomorphic\n20:52269701:C:A\tMonomorphic\n20:54179856:G:T\tMonomorphic\n20:59363320:G:A\tMonomorphic"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-06-15T15:37:22.79833"
    }
}